crates/worker/src/partition_processor_manager.rs

```diff
@@ -917,39 +928,53 @@ impl PartitionProcessorManager {
             .processor_states
             .iter()
+            .filter(|(partition_id, _)| !self.pending_snapshots.contains_key(partition_id))
             .filter_map(|(partition_id, state)| {
                 state
                     .partition_processor_status()
+                    .filter(|status| {
+                        status.effective_mode == RunMode::Leader
+                            && status.replay_status == ReplayStatus::Active
+                    })
                     .map(|status| (*partition_id, status))
+            });
+
+        let (mut known_archived_lsn, unknown_archived_lsn): (Vec<_>, Vec<_>) =
+            potential_snapshot_partitions.partition_map(|(partition_id, status)| {
+                match self.archived_lsns.get(&partition_id) {
+                    Some(&archived_lsn) => Either::Left((
+                        partition_id,
+                        status.last_applied_log_lsn.unwrap_or(Lsn::INVALID),
+                        archived_lsn,
+                    )),
+                    None => Either::Right(partition_id),
+                }
+            });
+
+        for partition_id in unknown_archived_lsn {
+            self.spawn_update_archived_lsn_task(partition_id, snapshot_repository.clone());
+        }
+
+        // Limit the number of snapshots we schedule automatically
+        const MAX_CONCURRENT_SNAPSHOTS: usize = 4;
+        let limit = MAX_CONCURRENT_SNAPSHOTS.saturating_sub(self.pending_snapshots.len());
+
+        known_archived_lsn.shuffle(&mut rand::rng());
+        let snapshot_partitions = known_archived_lsn
+            .into_iter()
+            .filter_map(|(partition_id, applied_lsn, archived_lsn)| {
+                if applied_lsn >= archived_lsn.add(Lsn::from(records_per_snapshot.get())) {
+                    Some(partition_id)
+                } else {
+                    None
+                }
             })
-            .filter(|(partition_id, status)| {
-                status.effective_mode == RunMode::Leader
-                    && status.replay_status == ReplayStatus::Active
-                    && status.last_applied_log_lsn.unwrap_or(Lsn::INVALID)
-                        >= self
-                            .archived_lsns
-                            .get(partition_id)
-                            .copied()
-                            .unwrap_or(Lsn::INVALID)
-                            .add(Lsn::from(records_per_snapshot.get()))
-            })
-            .collect();
-        snapshot_partitions.shuffle(&mut rand::rng());
-
-        for (partition_id, status) in snapshot_partitions.into_iter().take(limit) {
-            debug!(
-                %partition_id,
-                last_archived_lsn = %status.last_archived_log_lsn.unwrap_or(Lsn::OLDEST),
-                last_applied_lsn = %status.last_applied_log_lsn.unwrap_or(Lsn::INVALID),
-                "Creating periodic partition snapshot",
-            );
+            .take(limit);
```

Comment:
I understand that the shuffle+limit is to make sure we snapshot `limit` number partitions each time but make sure all partitions are snapshot overtime. But I am wondering if this is the best algorithm to guarantee the fairness, specially if a snapshoting is too fast, then same partition will become a candidate again for a snapshot on next iteration.

Won't it be better if we instead sort this list (DESC) on (applied_lsn - archived_lsn). It will also make sure hot partitions are snapshotted more often than others

> NOTE: This is just a suggestion, I have no proof that this is a better approach than what you have already implemented.

Reply:
This is a great suggestion as it optimizes for the most important metric: we want to minimize the amount of work done by a cold partition processor to start up. If some partitions are moving much faster than others, it's okay to bias towards them. I'd like to make it a mix of your suggestion + a degree of randomness. Since this was technically first introduced in #3060, let me rather do that as a separate follow up?